Controller : save metrics data to DB

kantera800 opened the pull request:
add support for combined infrastructure metrics flow in order to pass controller certification test 5.7.1
Task #209

1. Controller:  add handler for LINK_METRIC_RESPONCE_MESSAGE
2. database: parse and store Transmitter/Receiver link metric TLVs to database from all Agents  for later usage building COMBINED_INFRASTRUCTURE_METRICS_MESSAGE 
3. Controller: add handler for AP_METRICS_RESPONSE_MESSAGE
4. database: parse and store AP Metric Query TLV to database from all Agents  for later usage building COMBINED_INFRASTRUCTURE_METRICS_MESSAGE

Commit:
Controller : create link_metrics_data  map in db

Add new data class (link_metrics_data) under son::node.
Add map of this class to save tx/rx metric data from all agents.
Add method to database only for GW node (get_metric_data_map).

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -622,6 +622,13 @@ int db::get_node_last_ping_avg_ms(std::string mac)
     return n->last_ping_avg_ms;
 }
 
+std::list<sTxRxLinkMetricsData>& db::get_combined_metric_data_list(std::string mac)
```
Return value should be const.

We decided to pass std::string arguments as const-ref. However, we also decided to pass mac addresses as sMacAddr (and internally convert the to string as long as the database isn't updated internally).

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -622,6 +622,13 @@ int db::get_node_last_ping_avg_ms(std::string mac)
+std::list<sTxRxLinkMetricsData>& db::get_combined_metric_data_list(std::string mac)
+{
+    auto n = get_node(mac);
+    //todo add protection for empty node how to return empty list ??
```
Since the return value is by reference, the only solution is to have a static const variable for it.

arnout commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -28,6 +30,22 @@ typedef struct {
+    sMacAddr mac_of_an_interface_in_the_neighbor_al;
+    std::list<ieee1905_1::tlvReceiverLinkMetric::sLinkMetricInfo> PairMetricInfo;    
+    
+} sTxRxLinkMetricsData;
```
Long term, I think we should decompose the link metric data into the information which is actually useful.

For now, however, we can stay pretty close to the TLV structures.
Therefore, I think it should be
```cpp
typedef struct {
    sMacAddr al_mac_of_the_device_that_reports;
    sMacAddr al_mac_of_the_device_that_is_reported;
   std::list<ieee1905_1::tlvTransmitterLinkMetric::sInterfacePairInfo> transmitterLinkMetrics;
   std::list<ieee1905_1::tlvReceiverLinkMetric::sInterfacePairInfo> receiverLinkMetrics;
} sLinkMetricsData;
```

Also, please use doxygen comments.

kantera800 replied:
sure, Its just a quick draft commit to present the general idea.

arnout commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -181,6 +199,7 @@ class node {
         std::unordered_map<int8_t, sVapElement> vaps_info;
     };
     std::shared_ptr<radio> hostap;
+    std::list<sTxRxLinkMetricsData> metrics_data_list;
```
Since you'll need to update existing link metrics when new information is received from an agent, it would be more appropriate to use a map. Or a set (since the key is contained in the structure) but then you'll need to add an equality operator that only matches the al_mac addresses. Or it can be a map and remove the al_mac addresses from the link metrics struct (when iterating over the map, you get the keys anyway).

I'm inclined to go for the map.

Note that in principle, there should be an entry in the map for each combination of agent nodes (and you can get the list of agents with `get_all_connected_ires()`).

kantera800 replied:
I thought we are going to clear the list each time combined metrics msg is sent . 
you are right using map is better.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -224,6 +225,8 @@ bool master_thread::handle_cmdu_1905_1_message(Socket *sd, ieee1905_1::CmduMessa
     case ieee1905_1::eMessageType::OPERATING_CHANNEL_REPORT_MESSAGE:
         return handle_cmdu_1905_operating_channel_report(sd, cmdu_rx);
+    case ieee1905_1::eMessageType::LINK_METRIC_RESPONSE_MESSAGE:
+        return handle_cmdu_1905_link_metric_responce(sd, cmdu_rx);
```
It's response, not responce. Also in the commit message.

kantera800 replied:
:)

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -945,6 +948,72 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+    LOG(INFO) << "Received LINK_METRIC_RESPONSE_MESSAGE, mid=" << std::dec << int(mid);
+
+    // build metric responce message
+    if (!cmdu_tx.create(0, ieee1905_1::eMessageType::LINK_METRIC_RESPONSE_MESSAGE)) {
```
I guess this is a copy-paste mistake? You don't need to construct a response here.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -945,6 +948,72 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                return false;
+            }
+
+            //TODO: save tlvTransmitterLinkMetric parameters to database (lCombindMetricData)
```
It will be as simple as db::add_transmitter_link_metric(al_mac_of_reporter, al_mac_of_neighbour, list<sInterfacePairInfo>). Constructing that list is still a bit of code though because the tlvf doesn't do it :-(.

Commit:
Controller : save LINK_METRIC_RESPONSE_MESSAGE TLVs to DB

Add handler for CMDU.
Pull tx metric data from tlvTransmitterLinkMetric.
Pull tx metric data from tlvReceiverLinkMetric.
Save tx/rx metric data to map (mMetricData) under al_mac key.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -622,6 +622,17 @@ int db::get_node_last_ping_avg_ms(std::string mac)
     return n->last_ping_avg_ms;
 }
 
+const std::map<std::string, son::node::link_metrics_data> db::get_metric_data_map(std::string mac)
```
My earlier comment about the return value needing to be const was in the assumption that it is returned by reference. If it is returned by value, there's no need for the const (it will anyway have no effect). But I guess returning by const reference is more appropriate.

As I mentioned before: we decided that string parameters should be passed by const reference. I still think you should use sMacAddr here (and convert to string in the call to get_node), but it's not terribly important.

So, the signature should be:

```
const std::map<std::string, son::node::link_metrics_data>& db::get_metric_data_map(const std::string& mac)
```

kantera800 replied:
I can use the suggested signature but without the "cosnt" part since i cannot insert to the map otherwise.
see thread regarding this issue:

https://stackoverflow.com/questions/10531010/const-references-in-stdvector-elements

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -622,6 +622,17 @@ int db::get_node_last_ping_avg_ms(std::string mac)
+    auto n = get_node(mac);
+    if (!n) {
+        LOG(WARNING) << __FUNCTION__ << " - node " << mac << " does not exist!";
+        const std::map<std::string, son::node::link_metrics_data> mDummy;
```
This should be static I think.

The `m` prefix is used for class members. That's not the case here, so it should just be `dummy`.

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -231,6 +231,12 @@ class db {
+    
+    /*
+    * returns reference to map that holds link metric data per Agent.
+    * use the following with caution.
```
This remark makes little sense unless you explain what the caution should be...

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -231,6 +231,12 @@ class db {
     void disconnected_slave_mac_queue_push(std::string node_mac);
     std::string node_to_string(std::string mac);
+    
+    /*
```
We decided to use doxygen comments, so

```
    /**
     * @brief Get the link metric data of a specific agent
     *
     * @param mac The AL-MAC address of the agent of which to get the link metric data.
     * @return reference to the map that holds the link metric data of agent @a mac.
     */
```

arnout commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,45 @@ bool node::set_type(beerocks::eType type_)
+            LOG(ERROR) << "add_transmitter_link_metric getting operating class entry has failed!";
+            return false;
+        }
+        auto &InterfacePairInfo = std::get<1>(info_tuple);
```
This isn't going to work... the interfacePairInfo is a thin object that refers to the buffer in which the CMDU was received. So as soon as a new CMDU is received, it's going to be overwritten with new data.

You'll need to copy everything into a new struct without underlying buffer.

It sucks that TLVF makes this difficult...

arnout commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -182,6 +186,26 @@ class node {
+    * Created empty in all nodes
+    * Used only in GW node (retreave from DB method get_metric_data_map)
+    */
+    std::map<std::string, son::node::link_metrics_data> mOfMetricData;
```
This is not correct: now you can have only one neighbour per agent. You need a list of link_metrics_data.

Since we get the transmit and receive metrics separately (in separate TLVs), it's probably easier to have a completely separate map of  `agent1 -> list<(agent2, vector<interface_pair_tx_link_metric>)>` and `agent1 -> list<(agent2, vector<interface_pair_rx_link_metric>)>`.

Also, I see no reason at all why `agent1` can't be specified with a sMacAddr instead of as a string.

kantera800 replied:
good point. changed to:
    map(transmitting agent sMacAddr ) of  a map(reported neighbor agent sMacAddr ) of link metric data (tx/rx)

arnout commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -182,6 +186,26 @@ class node {
+        ~link_metrics_data(){};
+        
+        sMacAddr al_mac_of_the_device_that_is_reported;
+        std::list<ieee1905_1::tlvTransmitterLinkMetric::sInterfacePairInfo> transmitterLinkMetrics;
```
Usually the list of interface pairs will contain just one element. Also, we never add or remove from the list, when we receive a linkMetricsTLV we always get the complete list. So a vector is more appropriate.

morantr replied:
I don't agree, every tlv that will be received must be added to the database. Therefore vector doesn't fit here, maybe std::queue is a better idea, but list is good also.

kantera800 replied:
Yes vector is good enough, list its just its an overkill for data that is not going to be changed. @morantr all TLVs will be saved with both cases no issue.

arnout commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,45 @@ bool node::set_type(beerocks::eType type_)
+    return true;
+}
+
+bool node::link_metrics_data::add_reciever_link_metric(std::shared_ptr<ieee1905_1::tlvReceiverLinkMetric> RxLinkMetricData)
```
reciever -> receiver

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,45 @@ bool node::set_type(beerocks::eType type_)
+    int info_size = (int)TxLinkMetricData->interface_pair_info_length();
+    LOG(DEBUG) << "interface_pair_info_length =" << info_size;
+    for (int i = 0 ; i < info_size ; i++)
+    {
```
clearly you didn't run clang-format if the for loop bracket has a separate line.

morantr commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -182,6 +186,26 @@ class node {
+        link_metrics_data(){};
+        ~link_metrics_data(){};
+        
+        sMacAddr al_mac_of_the_device_that_is_reported;
```
The variable name is too long IMO.
Please change it to "al_mac" and do the explaining in a comment.

morantr commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -182,6 +186,26 @@ class node {
+    /*
+    * This map holds link metric data per Agent.
+    * Created empty in all nodes
+    * Used only in GW node (retreave from DB method get_metric_data_map)
```
retreave  -> retrieved.
But I think you better delete what inside the bracket and "Created empty in all nodes".
Also, why this is used only on GW node? The data should be saved on all bridge node (TYPE_GW/TYPE_IRE).
Please update the commit message accordingly.

morantr commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -182,6 +186,26 @@ class node {
+    * Created empty in all nodes
+    * Used only in GW node (retreave from DB method get_metric_data_map)
+    */
+    std::map<std::string, son::node::link_metrics_data> mOfMetricData;
```
You don't need to explicitly write `son::node::`, because you are already in this namespace.
Second, the coding camel case is not our coding convention, and the 'm' prefix does usually means "member" and not "map" (I guess this is the reason for "Of" inside "mOfMetricData")
It should have been, m_metric_data. But after you will change it to lists as Arnout proposed, use the same naming convention.

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -231,6 +231,12 @@ class db {
+    * returns reference to map that holds link metric data per Agent.
+    * use the following with caution.
+    */
+    const std::map<std::string, son::node::link_metrics_data> get_metric_data_map(std::string mac);
```
Please make the string argument const reference.

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,45 @@ bool node::set_type(beerocks::eType type_)
+
+bool node::link_metrics_data::add_transmitter_link_metric(std::shared_ptr<ieee1905_1::tlvTransmitterLinkMetric> TxLinkMetricData)
+{
+    al_mac_of_the_device_that_is_reported = TxLinkMetricData->al_mac_of_the_neighbor_whose_link_metric_is_reported_in_this_tlv();
```
Long names, please shorten them, on the yaml as well.

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,45 @@ bool node::set_type(beerocks::eType type_)
+bool node::link_metrics_data::add_transmitter_link_metric(std::shared_ptr<ieee1905_1::tlvTransmitterLinkMetric> TxLinkMetricData)
+{
+    al_mac_of_the_device_that_is_reported = TxLinkMetricData->al_mac_of_the_neighbor_whose_link_metric_is_reported_in_this_tlv();
+    int info_size = (int)TxLinkMetricData->interface_pair_info_length();
```
why is has to be int? It is q size and therefore unsigned.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -948,6 +950,89 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+    LOG(INFO) << "Received LINK_METRIC_RESPONSE_MESSAGE, mid=" << std::dec << int(mid);
+
+    //database reference for data starage
+    std::map<std::string, son::node::link_metrics_data> mMetricData = database.get_metric_data_map(database.get_gw_mac());
```
Use auto&, it will save writing such a long type, and the reference is important since you don't want it to be a copy, you want to insert new data inside it.

kantera800 replied:
ok

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -948,6 +950,89 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+    LOG(INFO) << "Received LINK_METRIC_RESPONSE_MESSAGE, mid=" << std::dec << int(mid);
+
+    //database reference for data starage
+    std::map<std::string, son::node::link_metrics_data> mMetricData = database.get_metric_data_map(database.get_gw_mac());
```
Why database.get_gw_mac()? You need to take the al_mac of the sender from the *sd.

kantera800 replied:
mMetricData  needs to be  saved in one place in DB.
You suggested under gw node.
so i'm using get_metric_data_map method just to get the reference to the map from that node.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -948,6 +950,89 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+            
+            //fill tx data from TLV
+            if(!NewMetricData.add_transmitter_link_metric(TxLinkMetricData))
+            {
```
You didn't run clang-format.sh here too.
we use
```cpp
if (cond) {
   // code
}
```
not 
```cpp
if (cond) 
{
   // code
}
```

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -948,6 +950,89 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+
+            // both tx/rx TLVs need to hold the same al_mac
+            std::string next_tlv_al_mac_str = network_utils::mac_to_string(RxLinkMetricData->al_mac_of_the_device_that_transmits());
+            if (al_mac.compare(next_tlv_al_mac_str) != 0) {
```
It is possible to compare strings that way, but `string1 == string2` is more readable IMO

kantera800 replied:
now that i use sMacAddr  i added network_utils::mac_compare

morantr commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -138,6 +138,16 @@ void network_utils::mac_from_string(uint8_t *buf, const std::string &mac)
 }
 
+bool network_utils::mac_compare(const sMacAddr &mac1, const sMacAddr &mac2)
+{
```
Please do it as operator `==` on sMacAddr.

kantera800 replied:
apparently this operator already exists .I'll use it.  commit removed.

morantr commented on common/beerocks/bcl/include/beerocks/bcl/network/net_struct.h:
```diff
@@ -22,6 +22,19 @@ inline bool operator==(sMacAddr const &lhs, sMacAddr const &rhs)
 
 inline bool operator!=(sMacAddr const &lhs, sMacAddr const &rhs) { return !(rhs == lhs); }
 
+inline bool operator<(const sMacAddr &m1, const sMacAddr &m2)
```
This is not one line function so it definitely should not be an inline function.

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -235,6 +235,14 @@ class db {
+     * @param mac The mac address of the db node that we use to save link metric data.
+     * @return reference to the map that holds link metric data of all agents.
+     */
+    std::map<sMacAddr, std::map<sMacAddr, son::node::link_metrics_data>> &get_link_metric_data_map(const std::string& mac);
```
Why did you decide to use `std::map` rather than `std::unordered_map`? I don't see how the fact that the data is ordered is helpful so reorder operation on every insertion could be saved.
I think that if you change it to unrodered_map you might not need the commit which you implement on the `<` operator.

kantera800 replied:
you are correct unordered_map doesn't need < operator.
currently we are just passing messages to pass certification so I guess unordered_map is fine.
@arnout @moran hope you don't change your mind tomorrow changing back to a std::list  :)

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -235,6 +235,14 @@ class db {
+    /**
+     * @brief Get the link metric data of a specific agent
+     *
+     * @param mac The mac address of the db node that we use to save link metric data.
```
Two notes:
1. We shouldn't use the word "we" in doxygen comments. Instead, you could rephrase it like the following and it will look better:
The mac address of the db node that **is used** to save link metric data.
2. It is best if you add param direction to the comments. For example:
_@param[in] mac The mac...._

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,47 @@ bool node::set_type(beerocks::eType type_)
+            LOG(ERROR) << "add_transmitter_link_metric getting operating class entry has failed!";
+            return false;
+        }
+        auto InterfacePairInfo = std::get<1>(info_tuple);
```
you need to use `auto&` that way you will not create a copy.

kantera800 replied:
@arnout  already commented on this , if i use reference then next cmdu will overwrite the content since it uses the same buffer.

morantr replied:
I'm not sure I understand.
InterfacePairInfo is a local variable in this function which holds data on the buffer, and its data is already being copied when you push it to the container.
@arnout can you please explain?

kantera800 replied:
since push_back stores data by value i guess you are right . will use reference.

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,47 @@ bool node::set_type(beerocks::eType type_)
+    size_t info_size = TxLinkMetricData->interface_pair_info_length();
+
+    for (size_t i = 0;
+         i < info_size / sizeof(ieee1905_1::tlvTransmitterLinkMetric::sInterfacePairInfo); i++) {
```
This is wrong. `TxLinkMetricData->interface_pair_info_length()` returns the number of elements and not the bytes length.
Please change it to:
```cpp
for (size_t i = 0; i < info_size; i++) {
   // code
}
```

kantera800 replied:
I already tested it with  i < info_size;
it doesn't work .
interface_pair_info_length() returns # of bytes

morantr replied:
You are right, I can see now that on `interface_pair_info_length()` implementation, the function returns `num_of_element * sizeof(sInterfacePairInfo)`. It is not so clear to me why it does it, instead of returning num of elements, but oh well.
 Anyway, if it confused me, it will probably confuse more people, so I suggest to do make it clear with comments:
```cpp
//  interface_pair_info_length() returns the length in bytes (number of elements * sizeof(sInterfacePairInfo).
size_t info_length_bytes = TxLinkMetricData->interface_pair_info_length();
auto info_length_elements = info_length_bytes / sizeof(ieee1905_1::tlvTransmitterLinkMetric::sInterfacePairInfo);

for (size_t i = 0; i < info_length_elements ; i++) {
   // code
}
```

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,47 @@ bool node::set_type(beerocks::eType type_)
+    size_t info_size = RxLinkMetricData->interface_pair_info_length();
+
+    for (size_t i = 0;
+         i < info_size / sizeof(ieee1905_1::tlvReceiverLinkMetric::sInterfacePairInfo); i++) {
```
Same here. Compare `i` to info_size without dividing it.

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,47 @@ bool node::set_type(beerocks::eType type_)
+            LOG(ERROR) << "add_receiver_link_metric getting operating class entry has failed!";
+            return false;
+        }
+        auto InterfacePairInfo = std::get<1>(info_tuple);
```
Same here `auto&`.

morantr commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -183,6 +187,27 @@ class node {
+        ~link_metrics_data(){};
+        
+        std::vector<ieee1905_1::tlvTransmitterLinkMetric::sInterfacePairInfo> transmitterLinkMetrics;
+        std::vector<ieee1905_1::tlvReceiverLinkMetric::sInterfacePairInfo> receiverLinkMetrics;
```
Why did you decide to use `std::vector`? Because you insert each time one element which causes reallocation.
Second, did you implement some code that erases elements from the container? Because if you push data to it over time, and not erase, it could lead to out of memory error.
Consider to change it to std::queue which holds up only X amount of measurements,  or use std::list which holds time stamps for each sample, and implement piece of code that removes elements older than Y sec.

kantera800 replied:
see @arnout  old comment  on this, i implemented it first as a list and we decided that vector is enough since this data vector is not going to be modified.  after reading this data vector (to fill the combined_metrics TLVs)  the element that contains it will be erased, so i guess the ~link_metrics_data()will take care of the memory part.

morantr replied:
I looked at the comment, and I disagreed with it.
We should choose std::vector over std::list if it is important for us to access an element in O(1) on non-sequential access, meaning access vector[0] and right away to access vector[5] for example. As well as to have the option to grow.
What you did here is to push elements one at the time which causes memory reallocation.
You could have used the vector constructor which takes iterator to the beginning and iterator to the end to save reallocation. But still, if you will erase elements you will still cause to reallocations.
Anyway, the erasing of elements does not exist and should be added.
@arnout

kantera800 replied:
i choose vector because i agree with @arnout that most of the times there will be only one element to this vector and a list is an overkill.
If we will need the ability to erase only one element in the future i'll add it.

tomereli replied:
I agree that erasing elements should be added, or just using timestamps and keeping the latest X elements. Why is this resolved then?

arnout replied:
When link metrics for a neighbour are updated, it will always contain *all* interface pairs. Erasing elements is therefore never needed.

And usually, there's only one interface pair (if there are more, there is a loop in the network which usually is bad...). So a vector is more efficient than a list. We could have used an array because the size is know at creation time (cfr. info_size in the add functions).

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1122,6 +1124,177 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+    // holds mac address of the reporting agent, used as m_link_metric_data key
+    sMacAddr reporting_agent_al_mac, neighbor_al_mac;
+    reporting_agent_al_mac.struct_init();
+    neighbor_al_mac.struct_init();
```
There is no need to call struct_init(). It is meant for use in the tlv init function.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1122,6 +1124,177 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+    // parse all tlvs in cmdu and save ap metric data to database
+    while ((tlvType = cmdu_rx.getNextTlvType()) != int(ieee1905_1::eTlvType::TLV_END_OF_MESSAGE)) {
+
+        LOG(DEBUG) << "getNextTlvType";
```
Please remove this print. It is not useful.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1122,6 +1124,177 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+            reporting_agent_al_mac = TxLinkMetricData->al_mac_of_the_device_that_transmits();
+            neighbor_al_mac =
+                TxLinkMetricData
+                    ->al_mac_of_the_neighbor_whose_link_metric_is_reported_in_this_tlv();
```
please rename this variable name in yaml file. It is not reasonable to use such a long name.

kantera800 replied:
done

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1122,6 +1124,177 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+
+            if (!network_utils::mac_compare(
+                    reporting_agent_al_mac,
+                    RxLinkMetricData->al_mac_of_the_device_that_transmits())) {
```
Need to change in the yaml file the name of this variable as well. It is too long.

kantera800 replied:
done

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1122,6 +1124,177 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+            }
+        }
+
+        if (tlvType == int(ieee1905_1::eTlvType::TLV_RECEIVER_LINK_METRIC)) {
```
You should use `else if` since both conditions could no be true at the same iteration.

kantera800 replied:
ok

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1122,6 +1124,177 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                LOG(ERROR) << "adding RxLinkMetricData has failed";
+                return false;
+            }
+        }
```
It is a good idea to add here an `else` for unexpected tlv with an error print.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1122,6 +1124,177 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+
+    //save agent link metric data to database or modify existing.
+    if (m_link_metric_data.find(reporting_agent_al_mac) != m_link_metric_data.end()) {
+        if (m_link_metric_data[reporting_agent_al_mac].find(neighbor_al_mac) !=
```
`find()` method already returns an iterator to the element, so it is best to use what you find instead of look it up again with `[]` operator.

kantera800 replied:
second find is for the second map (its a map[reporter] of a map[neighbor])

morantr replied:
You didn't understand me. I'll try to explain with code:
```cpp
auto neigbour_map_it = m_link_metric_data.find(reporting_agent_al_mac;
if (neigbour_map_it = m_link_metric_data.find(reporting_agent_al_mac) != m_link_metric_data.end()) {
    if (neigbour_map_it->second.find(something))
```

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1122,6 +1124,177 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                   << network_utils::mac_to_string(reporting_agent_al_mac) << std::endl
+                   << " adding new report for neighbor al_mac ="
+                   << network_utils::mac_to_string(neighbor_al_mac) << std::endl;
+    }
```
Correct me if I am wrong, but AFAIU, what you are trying to do here is to update the container from the database with new data, and if it doesn't have a specific mac, add it. In that case, all this code here is useless since `[]` operator does already insert a new element if it doesn't exist. This means that you could replace all of this code with one line:
`link_metric_data[reporting_agent_al_mac][reporting_agent_al_mac] = new_link_metric_data;`

kantera800 replied:
1. I would like to know what really happens in case of a failure.
2. in the future i'm not sure that we would like to update the results in all cases, maybe we would like to analyze it first.

in any case i think that it doesn't hurt and the code is more readable.

morantr replied:
But there is no failure. It is always, insertion of a new element or updates one, and I don't see how it is interesting which one has occurred.
IMO, leaving code for uncertain future purpose is a bad idea. If it doesn't contribute anything it better be removed.

kantera800 replied:
You need to understand that there is more than writing a pretty code.
It has to be functional.
We are in the middle of certification testing with multiple golden units.
Our job is to identify vulnerable spots and add prints that will be helpful for debugging issues like M3 msg with Qualcom agent.
Also,  this is temporary code that will be replaced by the actual functionality in further stages.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1122,6 +1124,177 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+    LOG(INFO) << "Received LINK_METRIC_RESPONSE_MESSAGE, mid=" << std::dec << int(mid);
+
+    //getting reference for link metric data storage from db
+    auto &m_link_metric_data = database.get_link_metric_data_map(database.get_gw_mac());
```
`m_` prefix is saved for class members, and should not be used to declare local variables.

kantera800 replied:
fixed

Commit:
Framework: rename long names in TLV parameters

Affected TLVs:
tlvTransmitterLinkMetric
tlvReceiverLinkMetric

preperation commit for save LINK_METRIC_RESPONSE_MESSAGE TLVs to DB.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

morantr commented:
LGTM so far.

Commit:
tlvf: Add TLV_AP_METRIC

Preparative commit for  save AP_METRICS_RESPONSE_MESSAGE TLVs to db.
Added auto generated code for TLV_AP_METRIC.
Important note:
Tlvf doesn’t support optional parameters.
Temp solution applied to pass certification.
Following optional parameters affected:
 estimated_service_info_field_ac_be:
 estimated_service_info_field_ac_bk:
 estimated_service_info_field_ac_vo:
 estimated_service_info_field_ac_vi:

W/A: define max size instead of dynamic.
Opened task #471 regarding this issues.

These parameters need to be reverted to dynamic (task  #473)

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
Controller : create ap_metrics_data map in db

Add new data class (ap_metrics_data) under son::node.
Add map of this class for saving metric data from all agents.
Add method to database only for GW node (get_ap_data_map).

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
Controller : save AP_METRICS_RESPONSE_MESSAGE TLVs to DB

Add handler for CMDU.
Read metric data from tlvApMetric.
Save metric data to map (m_ap_metric_data) under bssid key.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -343,3 +343,30 @@ bool node::link_metrics_data::add_receiver_link_metric(
+        if (ApMetricData->estimated_service_parameters().include_ac_be)
+        {
+            estimated_service_info_field_ac_be[i] = *ApMetricData->estimated_service_info_field_ac_be(i);
+        }
```
Please run clang-format

morantr commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -343,3 +343,30 @@ bool node::link_metrics_data::add_receiver_link_metric(
+        }
+    }
+    return true;
+}
```
missing empty line on the last line.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1160,6 +1163,39 @@ static void print_link_metric_map(
+        for (int i = 0; i < 3; i++) {
+            LOG(DEBUG) << std::hex << pair_agent.second.estimated_service_info_field_ac_bk[i];
+        }
+
```
It won't be printed nicely. I suggest to do it that way:
```cpp
LOG(DEBUG) << std::hex  <<
    "estimated_service_info_field_ac_be = 0x" << 
    int(pair_agent.second.estimated_service_info_field_ac_be[0]) <<
    int(pair_agent.second.estimated_service_info_field_ac_be[1]) <<
    int(pair_agent.second.estimated_service_info_field_ac_be[2]) << std::endl <<
    estimated_service_info_field_ac_bk = 0x" <<
    int(pair_agent.second.estimated_service_info_field_ac_bk[0]) <<
    int(pair_agent.second.estimated_service_info_field_ac_bk[1]) <<
    int(pair_agent.second.estimated_service_info_field_ac_bk[2]) << std::endl <<
    // other fields
    ;
```

kantera800 replied:
fixed

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -237,6 +237,15 @@ class db {
+     * @return reference to the map that holds link metric data of all agents.
+     */
+    std::unordered_map<sMacAddr, std::unordered_map<sMacAddr, son::node::link_metrics_data>> &
+    get_link_metric_data_map(const std::string &mac);
```
Please use sMacAddr for the `mac` argument as well.

CoralMalachi replied:
Do you want to define another method with sMacAddr as an argument (with the same method name) or to replace this one completely?

arnout replied:
On second thought, as mentioned elsewhere, the argument is useless because it is always the GW node.

Instead, put this map in the db directly, not in the node, and no need for an argument in this function.

arnout commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -183,6 +187,30 @@ class node {
+    /*
+    * This map holds link metric "data struct" per reporting Agent sMacAddr .
+    * "data struct" holds map of the actual link_metrics_data vector (tx/rx) per reported Agent sMacAddr.
+    * Map is Used in TYPE_GW/TYPE_IRE nodes.
```
This is not true I think: it is only used in the gateway node.

So I think it doesn't really make sense to have this map in the node. Just add it directly to the database.

kantera800 replied:
correct IRE is not relevant here, I'll remove it from comment .
I use method  in TYPE_GW anyway.
regarding saving it on the node, it was @morantr   suggestion.

morantr replied:
Currently, all measurements are being saved on the node that measured them.
I would keep it that way.

arnout replied:
> Currently, all measurements are being saved on the node that measured them.

In this PR, that statement is not true. All measurements are saved on the gateway node, not the node that made the measurements.

It would indeed be a good idea to store the measurements in the IRE (or GW) node that made the measurements. But then, we should have just a `std::unordered_map<sMacAddr, son::node::link_metrics_data>` in the node.

It would also mean that to construct the combined infrastructure metrics, we'd have to iterate over all nodes instead of just querying this one map.

kantera800 replied:
there is no point in saving metrics on the IRE at least not for certification.
I think what @morantr meant to say is that if  GW (controller) will be using  metric data measurements then  it should be save on its node's db.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1133,159 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+        LOG(DEBUG) << "sent from al_mac= " << network_utils::mac_to_string(pair_agent.first)
+                   << std::endl;
+
+        for (auto const &pair_neighbor : link_metric_data[pair_agent.first]) {
```
Doesn't this make more sense?
```
        for (auto const &pair_neighbor : pair_agent.second) {
```

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1133,159 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+{
+    LOG(DEBUG) << "Printing Link Metrics data map";
+    for (auto const &pair_agent : link_metric_data) {
+        LOG(DEBUG) << "sent from al_mac= " << network_utils::mac_to_string(pair_agent.first)
```
It would be useful to add indentation here so the structure is clear. So one space before "sent from al_mac", two spaces before "reporting neighbor" etc.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1133,159 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+
+            auto &vrx = pair_neighbor.second.receiverLinkMetrics;
+            for (unsigned int i = 0; i < vrx.size(); ++i) {
+                LOG(DEBUG) << "rx interface metric data # " << i << "  neighbor al_mac:"
```
This is not an al_mac but an interface MAC. Same below.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1133,159 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                           << network_utils::mac_to_string(vrx[i].neighbor_interface_mac)
+                           << "  receiving al_mac:"
+                           << network_utils::mac_to_string(vrx[i].rc_interface_mac)
+                           << "  rssi= " << std::hex << int(vrx[i].link_metric_info.rssi_db)
```
Why only report the rssi and not the other 3 fields?

kantera800 replied:
rssi will probably change so its good enough.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1133,159 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                           << " and TLV_TRANSMITTER_LINK_METRIC reporter al_mac ="
+                           << network_utils::mac_to_string(RxLinkMetricData->reporter_al_mac())
+                           << std::endl
+                           << " not the same";
```
There is no guarantee that Tx and Rx appear in any particular order, or that there is even a Tx. Therefore, I think both Tx and Rx should have this check, but only if `reporting_agent_al_mac != ZERO_MAC` (i.e. not for the first one).

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1133,159 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+            return false;
+        }
+    }
+
```
You're still missing an important case here: a single CMDU can contain several Rx and Tx link metrics for several neighbours. So if there are agents A, B and C, agent A can send link metrics with the following TLVs: `Tx<A,B>, Rx<A,B>, Rx<A,C>, Tx<A,C>`.  I intentionally swapped Rx and Tx in the middle to show that the order is not specified. What we can assume, however, is that if A reports link metrics, it will report them for all neighbours (if we were to ask for the link metrics of only one specific neighbour, we wouldn't get all of them, but for now we can assume we always ask for all neighbours - make sure there's a comment that explains this assumption).

Another problem is that this code only adds data, it doesn't remove stale neighbours. When a neighbour disappears, the reporting agent will no longer report link metrics for that neighbour. In that case, the neighbour should also be removed from the database.

We can handle both in one shot with the following pseudocode:

```
reporting_al_mac = <get it from the first TLV>
link_metric_data[reporting_al_mac].clear();
for (tlv) {
    if (tlv->reporting_al_mac != reporting_al_mac) {
       ERROR
    }
    link_metric_data[reporting_al_mac][neighbour_al_mac].set_from_tlv(tlv);
}
```

It gets complicated a little due to Rx and Tx being different classes, but you get the drift.

kantera800 replied:
not sure why do we need the assumption , the  map<map<link_metrics>> solution supports all combination. no sure what is so special in a case where agent asked not to report all his neighbors (we would add it to the report anyways).

For the second part , you are talking about the actual implementation.
I'm don't think we need to cover disappearing agent scenarios for certification.
but adding clear() method is not a bad idea .
So basically you prefer clearing all reporting agent's map every time we get new link metric from him.

p.s
i will  erase all link/ap metric db  after combined_metrics message is sent . please correct me if i'm wrong.

arnout replied:
> not sure why do we need the assumption , the map<map<link_metrics>> solution supports all combination. no sure what is so special in a case where agent asked not to report all his neighbors (we would add it to the report anyways).

Sorry, that statement referred to the removal of stale neighbours. To be able to remove stale neighbours, we have to know *when* a neighbour is stale. Since we always ask for all metrics, not for a specific neighbour, we know that when we get the metrics, it's for all neighbours. Therefore, we can simply clear all the existing metrics for that particular reporting agent.


> So basically you prefer clearing all reporting agent's map every time we get new link metric from him.

Yep.


> i will erase all link/ap metric db after combined_metrics message is sent . please correct me if i'm wrong.

That is wrong. In general, it is possible that you receive new metrics for one agent but not another, so you should only clear the old metrics when new metrics are actually received.

kantera800 replied:
ok.
I'll add the clear()  method and use it each time new metrics arrives (need to run the test again to see that it holds though).
will add some comments to explain this in the code.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1133,159 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+        } else {
+            neighbor_status_s = "existing";
+        }
+    }
```
This is a lot of code to just say whether it is the first or second time we receive link metrics from that particular reporter, so I wouldn't bother with it.

kantera800 replied:
i wasn't aware of the fact that new link metrics erases the old report (even if its for a specific neighbor)
Are you sure that is  the case  ?

arnout replied:
> i wasn't aware of the fact that new link metrics erases the old report (even if its for a specific neighbor)
> Are you sure that is the case ?

If it weren't, then after a couple of days you'd get a huge collection of metrics...

kantera800 replied:
ok, I'll remove it .

arnout commented on framework/tlvf/yaml/tlvf/wfa_map/tlvApMetric.yaml:
```diff
@@ -9,22 +9,22 @@ tlvApMetric:
   estimated_service_info_field_ac_be:
     _type: uint8_t
-    _length: [ ]
+    _length: [3]
```
As a permanent solution for the problem of the optional fields, you could define just this one field and make it dynamic:

```yaml
  estimated_service_info:
    _type: uint8_t
    _length: [ ]
```

The receiver can validate if the length corresponds to what is expected based on the estimated_service_parameters.

kantera800 replied:
the problem that it doesn't pass compilation.

arnout replied:
> the problem that it doesn't pass compilation.

It builds for me...

CoralMalachi replied:
It does not compile for me also @arnout, maybe you meant to define a length variable as defined here?

```yaml
cRestrictedOperatingClasses:
  _type: class
  operating_class: uint8_t
  channel_list_length:
    _type: uint8_t
    _length_var: True
  channel_list:
    _type: sChannelInfo
    _length: [ channel_list_length ]
```

arnout replied:
> It does not compile for me also @arnout, maybe you meant to define a length variable as defined here?

There is no length variable that can be used in the AP_METRICS structure, only the overall length of the TLV.

I did try it with exactly the code I wrote:
```
  estimated_service_info:
    _type: uint8_t
    _length: [ ]
```

and it built successfully (after, of course, adapting all the uses of estimated_service_info_*).

tomereli replied:
@CoralMalachi , @kantera800 you both didn't understand @arnout 's suggestion.
Instead of having 4 different unknown length lists which are optional based on the `estimated_service_parameters`, he suggested to replace it with a single unknown length list.
The failure you originally saw was that TLVF cannot accommodate more then one unknown length list OR an unknown length list which is not the *last* member in the TLV.
The reason for that is due to the fact that on parsing, we use the TLV `length` field and subtract all the rest of the class parameter sizes from it to get the unknown length list actual length.

So - instead of having 4 different unknown length lists which by now you should understand why it is not supported, lets define a single one:

```yaml
tlvApMetric:
  _type: class
  _is_tlv_class : True
  type:
    _type: eTlvTypeMap
    _value_const: TLV_AP_METRIC 
  length: uint16_t
  bssid: sMacAddr
  channel_utilization: uint8_t
  number_of_stas_currently_associated: uint16_t
  estimated_service_parameters: sEstimatedService
  estimated_service_info:
    _type: uint8_t
    _length: []
```

Usage:
First, understand the length of the info based on `sEstimatedService`, lets say it is `len`.
Then we simply allocate this list:
```
ap_metrics->alloc_estimated_service_info(len);
auto info = ap_metrics->estimated_service_info(); //get the buffer
```
Now we can access the buffer and modify it as we need based on `sEstimatedService`. We will also need to take care ourselves for any endianness conversion. 

However - this PR is blocked for too long and I really want to get it merged, and there is already a task for this - #473 - so let's keep the WA as is since by miracle it passes the test and @CoralMalachi will fix very soon as part of a separate PR.

arnout commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -211,6 +212,29 @@ class node {
+        uint8_t estimated_service_info_field_ac_vo[3];
+        uint8_t estimated_service_info_field_ac_vi[3];
+
+        bool add_ap_metric_data(std::shared_ptr<wfa_map::tlvApMetric> ApMetricData);
```
I don't really like that the wfa_map definitions leak through into the database. I would prefer if the add_ap_metric_data function was in the code that deals with wfa_map stuff - in other words, in the master. Admittedly, that file is already way too large, but we can split it up.

Note that this is just my opinion based on a very vague feeling of appropriateness, so you shouldn't necessarily take it into account.

arnout commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -211,6 +212,29 @@ class node {
+    * Map is Used in TYPE_GW/TYPE_IRE nodes.
+    * Map created empty in all other nodes.
+    */
+    std::unordered_map<sMacAddr, ap_metrics_data> m_ap_metric_data;
```
This map shouldn't be part of the GW/IRE node. Instead, it should be part of the node that corresponds to a BSS. Unfortunately, the db doesn't currently model it that way - we do have son::node::radio but no per-BSS (i.e. per-VAP) info. Actually, that is probably a mistake, because there is already the stats_info which contains the per-BSS stats.

In fact, probably you should just update ap_stats_params with the service info. active_sta_count and channel_load_percent are already there.

Alternatively, I'm OK with adding an issue to clean this up later.

arnout commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -343,3 +343,30 @@ bool node::link_metrics_data::add_receiver_link_metric(
+    channel_utilization                 = ApMetricData->channel_utilization();
+    number_of_stas_currently_associated = ApMetricData->number_of_stas_currently_associated();
+
+    for (int i = 0; i < 3; i++) {
```
It would feel much more natural to me to have the loop inside the conditions. Yes, that means we need to repeat the loop four times, but I think it's more appropriate.

Also we could use `std::copy(ApMetricData->estimated_service_info_field_ac_be(), ApMetricData->estimated_service_info_field_ac_be() + 3, estimated_service_info_field_ac_be)`

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1286,6 +1320,76 @@ bool master_thread::handle_cmdu_1905_link_metric_response(Socket *sd,
+    auto &ap_metric_data = database.get_ap_metric_data_map(database.get_gw_mac());
+
+    //will hold new ap metric data from Reporting Agent
+    son::node::ap_metrics_data new_ap_metric_data;
```
This can be moved inside the loop.

But actually, I don't even think it's needed, we can set it directly in the map without a temporary variable.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1286,6 +1320,76 @@ bool master_thread::handle_cmdu_1905_link_metric_response(Socket *sd,
+
+    int tlvType;
+    // holds mac address of the reporting agent, used as ap_metric_data key
+    sMacAddr reporting_agent_bssid;
```
This can be moved inside the loop. Also, the reporting_agent part is not really relevant IMO, it's just the BSSID of the BSS of which the statistics are reported.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1286,6 +1320,76 @@ bool master_thread::handle_cmdu_1905_link_metric_response(Socket *sd,
+
+        if (tlvType == int(wfa_map::eTlvTypeMap::TLV_AP_METRIC)) {
+            //parse tx_ap_metric_data
+            auto apMetricData = cmdu_rx.addClass<wfa_map::tlvApMetric>();
```
apMetricData and ap_metric_data are two very similar names for two completely different things....

arnout reviewed (changes requested):
I would also already add a skeleton for the combined infra metrics test to test_flows, and just checking the output of the print functions in the gateway log file. Otherwise, we're adding a lot of code here without any test.

Commit:
controller: send combined metrics cmdu

According to Test Plan, when CTT Agent 2 sends a 1905 Link Metric Response
message to the MCUT,The MCUT is triggered to send a combined
Infrastructure metrics message to CTT Agent 1.

For that reason, the following changes were made:

1. Agent subscribes to CMDU of type COMBINED_INFRASTRUCTURE_METRICS_MESSAGE.

2. After verifying that all agents sent Link Metric Response message to
   the controller, the controller creates and sends a combined
   Infrastructure metrics message to CTT Agent 1.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

tomereli reviewed (commented):
My major comment is on tlvf: Add TLV_AP_METRIC commit is that I completely agree with @arnout about the way to solve this, and both @CoralMalachi and @kantera800 apparently didn't understand. So I added an explanation and made sure it passes tlvf.py creation (tip - you can run only tlvf.py on the framework and see if it fails or passes using the prplMesh tlvf debug configuration).

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                return false;
+            }
+
+            if (reporting_agent_al_mac != network_utils::ZERO_MAC &&
```
The same check must also be done in the transmitter case.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                               << " not the same";
+                    return false;
+                }
+                if (neighbor_al_mac != RxLinkMetricData->neighbor_al_mac()) {
```
This is wrong. It is possible that the same link metrics CMDU contains several Rx and Tx link metrics TLVs for different neighbours. The only constraint is that the reporting agent should always be the same (it's the sender of the message). So this condition should be removed. And neighbor_al_mac doesn't need to be stored in a variable, it will only be used for a log message.

All this becomes a lot simpler with cmdu_rx.parse().

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                       << network_utils::mac_to_string(reporting_agent_al_mac) << std::endl
+                       << "reported  al_mac =" << network_utils::mac_to_string(neighbor_al_mac)
+                       << std::endl;
+
```
You need the same `if (!old_link_metrics_removed)` condition in this branch, because the order of Tx and Rx metrics is not fixed.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+               << std::endl;
+
+    print_link_metric_map(link_metric_data);
+
```
All of the stuff below should only be done in certification mode.

Actually, I'd like all of this to be put in a function that constructs the combined infra metrics. Doing it here is a terrible hack, the least we can do is to be able to reuse this code when it is really used.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+        database.fill_certification_tx_buffer(cmdu_tx);
+        return true;
+    }
+    return son_actions::send_cmdu_to_agent(sd, cmdu_tx);
```
Nothing should ever be sent as a response to a link metrics report, so we should just return true here. Well, we should never even get here because all of this last part should only be done in certification mode anyway.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+            std::string reporter_status_s = "new";
+            auto reporting_map_it         = ap_metric_data.find(reporting_agent_bssid);
+
+            if (reporting_map_it != ap_metric_data.end()) {
```
This makes no sense... You just added it a few lines above, so it will never be "new"! That shows that this piece of code that checks if it is new or not is completely useless.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                return false;
+            }
+
+            //save agent ap metric data to database or modify already existing.
```
It was saved just above this comment. The comment is anyway pretty useless, "add_ap_metric_data" should be obvious enough.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                reporter_status_s = "existing";
+            }
+
+            // ap_metric_data[reporting_agent_bssid] = new_ap_metric_data;
```
Please don't leave commented-out code.

arnout commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,89 @@ bool node::set_type(beerocks::eType type_)
+            std::copy(ApMetricData->estimated_service_info_field_ac_be(),
+                      ApMetricData->estimated_service_info_field_ac_be() + 3,
+                      estimated_service_info_field_ac_be);
+        }
```
This loop makes exactly the same copy 3 times! In other words, the loop is not needed.

arnout commented on controller/src/beerocks/master/db/node.cpp:
```diff
@@ -297,3 +297,89 @@ bool node::set_type(beerocks::eType type_)
 }
+
+bool node::link_metrics_data::add_transmitter_link_metric(
+    std::shared_ptr<ieee1905_1::tlvTransmitterLinkMetric> TxLinkMetricData)
```
The coding style says that we use lower_snake_case, so that should be `tx_link_metric_data`.

Some of the new code still uses CamelCase, but that's just to be consistent with the surrounding code.

Same for other new variables introduced in this PR.

I should have made this comment earlier, of course.

(nitpick)

arnout commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -184,6 +189,37 @@ class node {
+        ap_metrics_data(){};
+        ~ap_metrics_data(){};
+
+        sMacAddr bssid;
```
I'm not sure if it's useful to include bssid here, since it's part of the map already. But OK.

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -726,6 +739,22 @@ class db {
+    * This map holds ap metric data per reporting Agent sMacAddr .
+    * Map is Used in TYPE_GW/TYPE_IRE nodes.
+    * Map created empty in all other nodes.
+    */
```
Add a TODO that this should move to the BSS nodes (which currently don't exist) instead of being a separate map.

CoralMalachi replied:
Done in #498 .

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -726,6 +739,22 @@ class db {
+    * "data struct" holds map of the actual link_metrics_data vector (tx/rx) per reported Agent sMacAddr.
+    * Map is Used in TYPE_GW/TYPE_IRE nodes.
+    * Map created empty in all other nodes.
+    */
```
Add a TODO that this should move to the agent nodes instead of being a separate map.

CoralMalachi replied:
Done in #498 .

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+    sMacAddr reporting_agent_al_mac, neighbor_al_mac;
+
+    // parse all tlvs in cmdu and save ap metric data to database
+    while ((tlvType = cmdu_rx.getNextTlvType()) != int(ieee1905_1::eTlvType::TLV_END_OF_MESSAGE)) {
```
Now that we have `cmdu_rx.parse()`, we could use that instead. That also removes the need for old_link_metrics_removed.

It's a big change though, so nitpick.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                link_metric_rx_tlv->reporter_al_mac() = agent.first;
+                link_metric_rx_tlv->neighbor_al_mac() = per_neighbor.first;
+                for (auto &interface_pair_info : vrx) {
+                    LOG(DEBUG) << "     interface_pair_info_length BEFORE ALLOC:  "
```
This debug is not needed - it was probably added to check some issue while developing, but I don't think it's still useful.

CoralMalachi replied:
You are right I forgot to remove this.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                    auto interfeace_idx =
+                        link_metric_rx_tlv->interface_pair_info_length() /
+                        sizeof(ieee1905_1::tlvReceiverLinkMetric::sInterfacePairInfo);
+                    LOG(DEBUG) << "     interface_pair_info_length AFTER ALLOC:  "
```
Same here, not needed.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+
+    for (auto &agent : link_metric_data) {
+
+        for (auto &per_neighbor : link_metric_data[agent.first]) {
```
Isn't this just:

```suggestion
        for (auto &per_neighbor : agent.second) {
```

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1131,6 +1137,363 @@ bool master_thread::handle_cmdu_1905_channel_selection_response(Socket *sd,
+                        return false;
+                    }
+
+                    auto interfeace_idx =
```
This would be a lot simpler as:

```
if (!link_metric_rx_tlv->alloc_interface_pair_info(vrx.size())) {
    LOG(ERROR) << "alloc_interface_pair_info() has failed!";
    return false;
}
size_it interface_idx = 0;
for (auto &interface_pair_info : vrx) {
    auto interface_tuple =
        link_metric_rx_tlv->interface_pair_info(interface_idx);
    ...
    interface_idx++;
}
```

(btw, spelling mistake in interfeace_idx)

arnout reviewed (approved):
I still have a lot of comments, not all of them nitpicky.

Still, they can be resolved in follow-up PRs, so I'm approving already. If someone merges it, please also create an issue that refers to the open comments in this PR.

CoralMalachi commented:
> My major comment is on tlvf: Add TLV_AP_METRIC commit is that I completely agree with @arnout about the way to solve this, and both @CoralMalachi and @kantera800 apparently didn't understand. So I added an explanation and made sure it passes tlvf.py creation (tip - you can run only tlvf.py on the framework and see if it fails or passes using the prplMesh tlvf debug configuration).

#498
